feat(stdlib) Append children to XML parsed custom nodes

alimnios72 opened the pull request:
Added support for parsing XML component's children and their respective fields.

karthick-sh committed:
Remaining ifsgnodechildren functions (#6)

* Add implementations of replacechild, removechildren and removechildrenindex

* Comment and add tests for removechildren, replacehild and removechildrenindex

* Implement getchild, insertchild and appendchildren along with tests

* Add appendchildren, createchildren and inserchildren along with tests for all methods

* Implement removechildindex and reparent and add tests

* Code cleanup and comments

* Code cleanup and pr comment changes

* Add comments to replaceChildAtIndex

alimnios72 committed:
feat(stdlib) Append children to XML parsed custom nodes

underwoo16 commented on src/brsTypes/components/RoSGNode.ts:
```diff
@@ -767,3 +768,25 @@ function addFields(interpreter: Interpreter, node: RoSGNode, typeDef: ComponentD
+            let setField = newChild.getMethod("setfield");
+            for (let [key, value] of Object.entries(child.fields)) {
+                if (setField) {
+                    setField.call(interpreter, new BrsString(key), new BrsString(value));
```
Do we know that all of the fields will be `BrsString`?

alimnios72 replied:
The only other type I've seen in these fields is an array type (i.e. `translation="[1,20]"`) but there is no way to know the type unless all built-in Roku components are loaded. I think for now this should be fine until we know what we'll do with the built-in nodes

underwoo16 commented on src/componentprocessor/index.ts:
```diff
@@ -19,6 +19,12 @@ interface ComponentField {
 
+interface ComponentNode {
+    name: string;
+    fields: Object;
```
Is there a way we could use a more informative type than `Object` here?

alimnios72 replied:
It would be something like this:
```
interface ComponentField {
  [key: string]: [value: string];
}
```
but don't think this is pretty useful. I'll give a try later

underwoo16 reviewed (commented):
Looks like a great start, I have a couple concerns off the bat though. Will need to take another look.

alimnios72 commented:
@underwoo16 Btw, can you add Gin as a reviewer please?

alimnios72 committed:
Addressed feedback

TwitchBronBron committed:
fix(lex): Return boolean from Lexer.isToken (#286)

Updated lexer.isToken to actually return a boolean.

lvcabral committed:
Implemente global string function String() and StringI() #297 (#299)

lvcabral committed:
Implemented roDateTime object #129 (#277)

* Implemented ifArrayJoin interface for roArray #125

* Implemented Sort and Reverse methods from ifArraySort

* Implementing SortBy method (incomplete)

* Changed edge case of non string to return empty string instead of raising an error #125

* Fixed new test for joing with invalid array item

* Implemented SortBy method and changed error handling on Sort method

* Refactored sortBy and sort methods

* Implemented unit tests for sort and reverse methods

* Implemented test cases for SortBy method

* Fixed comparison operators (static analysis)

* Implemented roDateTime component

* Correctly grouped unit tests for Sort and SortBy methods

* Implemented unit tests for roDateTime

* Fixed JSDoc (static analysis)

* Fixed timezone test

* Removed timezone offset from file test (alterady covered on unit tests)

* Set the default locale for roDateTime as "en-US" (brs needs to address localization later)

lvcabral committed:
Fixed `for loop` issue that should not process loop when step is on the wrong direction of the start-end parameters #310 (#311)

* Fixed for loop issue #310

* Fixed "for loop" to correctly handle steps #315

TwitchBronBron committed:
fix(parse): Include surrounding whitespace in Block nodes(#257)

Updates `Stmt.Block` locations to include surrounding whitespace, and
renamed `startingLocation` to `location`. 

fixes #159

sjbarag committed:
fix(lint): Fix lint errors

lvcabral committed:
feat(cli):  Exit via "quit"/"exit", and remove spurious "invalid" messages (#290)

fixes #272 
fixes #273

sjbarag committed:
chore(test): Remove spurious console.log (#321)

Whoops!

sjbarag committed:
chore(deps): Upgrade all dependencies (#322)

Still gotta move to `eslint` over `tslint`, but that's a problem for
future-me.

Thanks, `yarn upgrade`!

sjbarag committed:
fix(env): Remove separate mock scope

Our extensions can live in global scope - no need to complicate things!

sjbarag committed:
Merge pull request #284 from underwoo16/master

Implement ifSGNodeChildren and ifSGNodeDict interfaces, As well as a
function for mocking components like so:
`_brs_.mockComponent("COMPONENT_NAME", {MOCK_OBJECT_TO_RETURN})`

[ifSGNodeChildren](https://developer.roku.com/en-gb/docs/references/brightscript/interfaces/ifsgnodechildren.md)
and
[ifSGNodeDict](https://developer.roku.com/en-gb/docs/references/brightscript/interfaces/ifsgnodedict.md)

lvcabral committed:
feat(parse,interp): Implement bitshift operations (#292)

fixes #213

sjbarag committed:
chore(naming): rename src/mocks/ -> src/extensions/ (#325)

It's going to contain more than just mocks pretty soon!

sjbarag committed:
feat(extension): Add _brs_.runInScope (#326)

While the standard `Run` function allows arbitrary files to be executed,
they're executed in their own empty scope!  That leads to a whole lot of
repeated lexing, parsing, etc. of the `source/` tree when running unit
tests, so we can expose a custom version of `Run` that keeps the
previous values in-scope, deletes the previous `main` function, and
otherwise executes the requested files the same way.

lvcabral committed:
fix(types,interp): Automatically convert between Int and Float (#291)

Roku does truncate float to integer (not round) so I also changed the
constructor of Int32 and Int64 to correctly mimic the behavior.

fixes #280

lvcabral committed:
fix(types): Return `true` for invalid values on `RoAssociativeArray#doesExist` (#323)

Implements `isEmpty()` for `RoArray` and `RoAssociativeArray` (from
`ifEnum` interface), and ensures `doesExist()` returns `true` for
explicitly `invalid` properties.

fixes #320

sjbarag committed:
v0.15.0 (#327)

alimnios72 committed:
feat(stdlib): Implement observeField method (#268)

This PR adds support for the `observeField` method defined in the
`ifSGNodeField` interface by executing every named function attached to
an object field.

WilliamsMartinezGLB committed:
feat(stdlib): Implement roBoolean, roDouble, roFloat, roInt (#294)

Implements the following components:

- roBoolean
- roDouble
- roFloat
- roInt

alimnios72 committed:
Merge branch 'master' into features/node_def_parsing

alimnios72 committed:
Merge master into features/node_def_parsing

alimnios72 committed:
Fixed createObject method

alimnios72 committed:
Removed console logs and unused imports

alimnios72 committed:
Merge branch 'features/node_def_parsing' into rosgnode-children

alimnios72 commented:
Will reopen against upstream
